lib/backend/watchersyncer/watchercache.go

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchersyncer
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+type watcherCache struct {
+	logger       *logrus.Entry
+	client       api.Client
+	watch        api.WatchInterface
+	resources    map[string]resourceKey
+	oldResources map[string]resourceKey
+	results      chan<- interface{}
+	hasSynced    bool
+	resourceType ResourceType
+}
+
+var (
+	ListRetryInterval = 1000 * time.Millisecond
+	WatchPollInterval = 5000 * time.Millisecond
+)
+
+// resourceKey is an entry in our cache.  It groups the a key with the last known
+// revision that we processed.
+type resourceKey struct {
+	revision string
+	key      model.Key
+}
+
+// Create a new watcherCache.
+// Note that the results channel is untyped.  The watcherSyncer expects one of the following
+// types:
+// -  An error
+// -  An api.Update
+// -  A api.SyncStatus (only for the very first InSync notification)
+func newWatcherCache(client api.Client, resourceType ResourceType, results chan<- interface{}) *watcherCache {
+	return &watcherCache{
+		logger:       logrus.WithField("ListRoot", model.ListOptionsToDefaultPathRoot(resourceType.ListInterface)),
```

Comment:
Wouldn't it be nicer to have this context logger log out the resource type rather than the etcd path?

Reply:
it could - but this will work for non "resource types" too - in particular I'll want this for the allocation blocks, which I'm still on the fence about resource-ifying.

Reply:
the path contains more useful info IMHO, but for the mainline - we could just print out the resource type....

yeah ok.   I'll print out the valueType....   that'll work for all key types.

Reply:
Actually that might be easier said than done without plumbing a bunch of other stuff in.

If you don't mind, I'll leave as is and raise a post-Beta fix up to change this.